### PR TITLE
Adding Channelwise FC module to Python Frontend

### PR DESCRIPTION
**Pull request by szaman19**
- Added support for channelwise FC to the python frontend modules
- Added non-square kernels, dilation, and padding for convolution modules

### CODE REVIEW

File: python/lbann/modules/base.py [hunk 1]
```diff
@@ -216,34 +309,35 @@ def __init__(self, num_dims,
...
+        if (self.has_vectors):
+          kwargs['conv_dims'] = self.kernel_size
+          kwargs['conv_pads'] = self.padding
+          kwargs['conv_dilations'] = self.dilation
+          kwargs['conv_strides'] = self.stride
+        else:
+          kwargs['conv_dims_i'] = self.kernel_size
+          kwargs['conv_pads_i'] = self.padding
+          kwargs['conv_dilations_i'] = self.dilation
+          kwargs['conv_strides_i'] = self.stride
```
Comment (timmoon10) [inline python/lbann/modules/base.py:334]: If we commit to using the `has_vectors` API internally:
```suggestion
        kwargs['has_vectors'] = True
        kwargs['conv_dims'] = str_list(self.kernel_dim)
        kwargs['conv_pads'] = str_ilst(self.padding)
        kwargs['conv_dilations'] = str_list(self.dilation)
        kwargs['conv_strides'] = str_list(self.stride)
```

**Review by timmoon10 — APPROVED** [review]
Mostly looks good, although I have some suggestions for a nicer API.

File: python/lbann/modules/base.py (hunk 1 above)
Comment (szaman19) [reply python/lbann/modules/base.py]: This is a better way to do it. Changed to this in latest commits.

File: python/lbann/modules/base.py [hunk 2]
```diff
@@ -129,6 +130,95 @@ def forward(self, x):
...
+
+  def __init__(self,
+               size,
+               bias=False,
+               weights=[],
+               activation=None,
+               transpose=False,
+               name=None,
+               data_layout='data_parallel',
+               parallel_strategy={}):
```
[the remark this answers is not in the document]
Comment (szaman19) [reply python/lbann/modules/base.py]: Updated in latest commit to remove these options.

File: python/lbann/modules/base.py [hunk 3]
```diff
@@ -129,6 +130,95 @@ def forward(self, x):
         else:
             return y
 
+class ChannelwiseFCModule(Module):
```
[the remark this answers is not in the document]
Comment (szaman19) [reply python/lbann/modules/base.py]: Updated the name on the latest commit to be more explicit.

File: python/lbann/modules/base.py [hunk 4]
```diff
@@ -320,18 +356,12 @@ def forward(self, x):
...
-          kwargs['conv_dims'] = self.kernel_size
-          kwargs['conv_pads'] = self.padding
-          kwargs['conv_dilations'] = self.dilation
-          kwargs['conv_strides'] = self.stride
-        else:
-          kwargs['conv_dims_i'] = self.kernel_size
-          kwargs['conv_pads_i'] = self.padding
-          kwargs['conv_dilations_i'] = self.dilation
-          kwargs['conv_strides_i'] = self.stride
+        kwargs['conv_dims_i'] = str_list(self.kernel_size)
```
Comment (timmoon10) [inline python/lbann/modules/base.py:358]:
```suggestion
        kwargs['conv_dims_i'] = str_list(self.kernel_dims)
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: python/lbann/modules/base.py [hunk 5]
```diff
@@ -129,6 +131,92 @@ def forward(self, x):
...
+    super().__init__()
+    ChannelwiseFullyConnectedModule.global_count += 1
+    self.instance = 0
+    self.size = size
+    self.bias = bias
+    self.transpose = transpose
+    self.name = (name
+                 if name
+                 else 'channelwisefc{0}'.format(ChannelwiseFullyConnectedModule.global_count))
+    self.parallel_strategy = 'data_parallel'
```
Comment (timmoon10) [inline python/lbann/modules/base.py:176]:
```suggestion
    self.data_layout = 'data_parallel'
```

File: python/lbann/modules/base.py [hunk 6]
```diff
@@ -129,6 +131,92 @@ def forward(self, x):
...
+    self.instance += 1
+    name = '{0}_instance{1}'.format(self.name, self.instance)
+    y = lbann.ChannelwiseFullyConnected(x,
+                                        weights=self.weights,
+                                        name=(name+'_fc' if self.activation else name),
+                                        data_layout=self.data_layout,
+                                        output_channel_dims=self.size,
+                                        bias=self.bias,
+                                        transpose=self.transpose,
+                                        parallel_strategy=self.parallel_strategy)
```
Comment (timmoon10) [inline python/lbann/modules/base.py:210]:
```suggestion
                                        transpose=self.transpose)
```

File: python/lbann/modules/base.py [hunk 7]
```diff
@@ -129,6 +131,92 @@ def forward(self, x):
...
+    if self.activation:
+        return self.activation(y,
+                               name=name+'_activation',
+                               data_layout=self.data_layout,
+                               parallel_strategy=self.parallel_strategy)
```
Comment (timmoon10) [inline python/lbann/modules/base.py:215]:
```suggestion
                               data_layout=self.data_layout)
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

**Comment by szaman19** [comment]
@timmoon10 Updated with the requested changes. LeNet integration test passing.

**Review by timmoon10 — APPROVED** [review]

**Comment by timmoon10** [comment]
Judging from the diff, I take it there was some Git mishap? Can you rebase on `develop` just so we can be sure we don't mess things up? If that looks daunting, just copy the changed files (only `python/lbann/modules/base.py`?) into a new branch and open a new PR.

**Comment by szaman19** [comment]
Oh boy, I didn't notice the diff. Yeah, no idea what happened there, but I definitely didn't touch 667 files haha. 

Should be fixed now!

**Pull request merged**